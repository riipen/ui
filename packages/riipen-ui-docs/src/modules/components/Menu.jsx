import React from "react";

import Link from "src/modules/components/Link";
import pages from "src/pages";

import ThemeContext from "@riipen-ui/styles/ThemeContext";

class Menu extends React.Component {
  static contextType = ThemeContext;

  render() {
    const theme = this.context;

    return (
      <div>
        <ul>
          {pages.map((parent, i) => (
            <li key={i}>
              {parent.name}
              <ul>
                {parent.children.map((child, j) => (
                  <li key={j}>
                    <Link href={child.pathname}>{child.name}</Link>
                  </li>
                ))}
              </ul>
            </li>
          ))}
        </ul>
        <style jsx>{`
          div {
            font-family: ${theme.typography.fontFamily};
            width: 250px;
          }
          ul {
            font-weight: ${theme.typography.fontWeight.bold};
            list-style-type: none;
            margin: 0;
            padding: 0;
          }
          ul ul {
            font-weight: ${theme.typography.fontWeight.regular};
            list-style-type: none;
            margin: 0;
            padding-left: 10px;
          }
          li {
            padding: ${theme.spacing(1)}px;
          }
        `}</style>
      </div>
    );
  }
}

export default Menu;
