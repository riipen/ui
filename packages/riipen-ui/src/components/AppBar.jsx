import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";

class AppBar extends React.Component {
  static propTypes = {
    /**
     * The content of the component.
     */
    children: PropTypes.node,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.object,

    /**
     * The color of the component. It supports those theme colors that make sense for this component.
     */
    color: PropTypes.oneOf(["primary", "secondary", "white", "black"]),

    /**
     * The positioning type. The behavior of the different options is described
     * [in the MDN web docs](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Positioning).
     * Note: `sticky` is not universally supported and will fall back to `static` when unavailable.
     */
    position: PropTypes.oneOf([
      "absolute",
      "fixed",
      "relative",
      "static",
      "sticky"
    ])
  };

  static defaultProps = {
    color: "primary",
    position: "fixed"
  };

  static contextType = ThemeContext;

  render() {
    const { children, classes, color, position } = this.props;

    const theme = this.context;

    const className = clsx(color, position, classes);

    return (
      <React.Fragment>
        <div className={className}>{children}</div>
        <style jsx>{`
          div {
            align-items: center;
            box-shadow: ${theme.shadows[3]};
            box-sizing: border-box;
            display: flex;
            padding: ${theme.spacing(3)}px;
            width: 100%;
            z-index: ${theme.zIndex.low};
          }

          .fixed {
            left: auto;
            position: fixed;
            right: 0;
            top: 0;
          }
          .absolute {
            left: auto;
            position: absolute;
            right: 0;
            top: 0;
          }
          .stickey {
            left: auto;
            position: sticky;
            right: 0;
            top: 0;
          }
          .static {
            position: static;
          }
          .relative {
            position: relative;
          }

          .primary {
            background-color: ${theme.palette.primary.main};
            color: ${theme.palette.primary.contrast};
          }
          .secondary {
            background-color: ${theme.palette.secondary.main};
            color: ${theme.palette.secondary.contrast};
          }
          .white {
            background-color: ${theme.palette.common.white};
            color: ${theme.palette.common.black};
          }
          .black {
            background-color: ${theme.palette.common.black};
            color: ${theme.palette.common.white};
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default AppBar;
