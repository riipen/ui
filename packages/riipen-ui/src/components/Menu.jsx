import React from "react";

import PropTypes from "prop-types";

import withClasses from "../utils/withClasses";

import MenuList from "./MenuList";
import Popover from "./Popover";

class Menu extends React.Component {
  static propTypes = {
    /**
     * The element the menu is attached too
     */
    anchorEl: PropTypes.object,

    /**
     * The location to attach the content too on the anchor element
     */
    anchorPosition: PropTypes.object,

    /**
     * Whether to autofocus the first element on first open
     */
    autoFocus: PropTypes.bool,

    /**
     * The content of the component.
     */
    children: PropTypes.node,

    /**
     * Array of additional CSS classes to use.
     */
    classes: PropTypes.array,

    /**
     * Whether or not the menu should close when an option is chosen with a click event
     */
    closeOnClick: PropTypes.bool,

    /**
     * The color of the component. It supports those theme colors that make sense for this component.
     */
    color: PropTypes.oneOf(["primary", "secondary"]),

    /**
     * The location to attach the anchor to on the content element
     */
    contentPosition: PropTypes.object,

    /**
     * Whether or not the menu should be rendered
     */
    isOpen: PropTypes.bool,

    /**
     * Whether or not the popout should be forced to stay on screen
     */
    keepOnScreen: PropTypes.bool,

    /**
     * The function callback for when the selection is changed
     */
    onChange: PropTypes.func,

    /**
     * The function callback to use when the menu closes
     */
    onClose: PropTypes.func,

    /**
     * The styles to be applied to the popover list
     */
    popoverStyles: PropTypes.object,

    /**
     * The index of the item selected in the list
     */
    selectedIndex: PropTypes.number,

    /**
     * The type of menu to create
     * Use 'menu' for lists of links
     */
    variant: PropTypes.oneOf(["menu", "selection"])
  };

  static defaultProps = {
    color: "primary",
    classes: [],
    closeOnClick: true,
    isOpen: true,
    keepOnScreen: false,
    popoverStyles: {},
    variant: "menu"
  };

  componentDidUpdate() {
    const { anchorEl } = this.props;
    if (document.activeElement && anchorEl) {
      document.activeElement.blur();
    }
  }

  handleChange = (idx, event) => {
    const { onChange, closeOnClick } = this.props;
    if (onChange) onChange(idx, event);
    if (closeOnClick && event && event.type === "click") this.handleClose(idx);
  };

  render() {
    const {
      anchorEl,
      anchorPosition,
      children,
      contentPosition,
      keepOnScreen,
      isOpen,
      onClose,
      popoverStyles,
      selectedIndex,
      variant
    } = this.props;

    return (
      <React.Fragment>
        <Popover
          onClose={onClose}
          anchorPosition={anchorPosition}
          contentPosition={contentPosition}
          lockScroll={false}
          isOpen={isOpen}
          anchorEl={anchorEl}
          styles={popoverStyles}
          keepOnScreen={keepOnScreen}
        >
          <MenuList
            {...this.props}
            selectChange={this.handleChange}
            selectedIndex={selectedIndex}
            variant={variant}
          >
            {children}
          </MenuList>
        </Popover>
      </React.Fragment>
    );
  }
}

export default withClasses(Menu);
