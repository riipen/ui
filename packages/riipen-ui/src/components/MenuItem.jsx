import React from "react";
import PropTypes from "prop-types";

import clsx from "clsx";
import css from "styled-jsx/css";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

import ListItem from "./ListItem";

class MenuItem extends React.Component {
  static propTypes = {
    /**
     * The content of the component.
     */
    children: PropTypes.node,

    /**
     * Array of additional CSS classes to use.
     */
    classes: PropTypes.array,

    /**
     * The color of the component. It supports those theme colors that make sense for this component.
     */
    color: PropTypes.oneOf(["primary", "secondary"]),

    /**
     * Whether or not the item is disabled
     */
    disabled: PropTypes.bool,

    /**
     * The function callback for when a menu item is selected
     */
    onClick: PropTypes.func,

    /**
     * Whether or not the item is selected
     */
    selected: PropTypes.bool
  };

  static defaultProps = {
    color: "primary",
    classes: []
  };

  getLinkedStyles = () => {
    const theme = this.context;

    return css.resolve`
      .menu-item {
        border-left: 5px solid transparent;
        box-sizing: border-box;
        color: ${theme.palette.text.secondary};
        cursor: pointer;
        font-family: ${theme.typography.body1.fontFamily};
        font-size: ${theme.typography.body1.fontSize};
        font-weight: ${theme.typography.body1.fontWeight};
        letter-spacing: ${theme.typography.body1.letterSpacing};
        line-height: ${theme.typography.body1.lineHeight};
        margin: 0;
        outline: 0;
        padding: ${theme.spacing(2)}px ${theme.spacing(4)}px;
        text-align: inherit;
      }

      .menu-item:hover {
        background-color: ${theme.palette.grey[100]};
      }

      .menu-item.disabled:hover {
        background-color: ${theme.palette.grey[50]};
        border-color: transparent;
        color: ${theme.palette.text.secondary};
      }

      .menu-item.primary:hover {
        border-color: ${theme.palette.primary.main};
      }

      .menu-item.secondary:hover {
        border-color: ${theme.palette.secondary.main};
      }

      .menu-item.selected {
        background-color: ${theme.palette.grey[200]};
        color: ${theme.palette.text.primary};
      }

      .menu-item.selected.primary {
        border-color: ${theme.palette.primary.main};
      }

      .menu-item.selected.secondary {
        border-color: ${theme.palette.secondary.main};
      }

      .menu-item.disabled {
        background-color: ${theme.palette.grey[50]};
        border-color: transparent;
        cursor: inherit;
        font-weight: ${theme.typography.fontWeight.light};
        opacity: 0.6;
      }

      .menu-item.disabled:hover {
        background-color: ${theme.palette.grey[50]};
        border-color: transparent;
      }
    `;
  };

  static contextType = ThemeContext;

  render() {
    const { onClick, classes, children, disabled, selected } = this.props;

    const linkedStyles = this.getLinkedStyles();

    const className = clsx(
      "menu-item",
      linkedStyles.className,
      {
        selected,
        disabled
      },
      classes
    );

    const handleClick = event => {
      if (disabled) {
        event.stopPropagation();
      }

      if (onClick && !disabled) onClick(event);
    };

    return (
      <React.Fragment>
        <ListItem {...this.props} onClick={handleClick} classes={[className]}>
          {children}
        </ListItem>
        {linkedStyles.styles}
      </React.Fragment>
    );
  }
}

export default withClasses()(MenuItem);
