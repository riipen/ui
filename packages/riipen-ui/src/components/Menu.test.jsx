import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";
import ReactTestUtils from "react-dom/test-utils";

import Menu from "./Menu";
import MenuItem from "./MenuItem";
import AppBar from "./AppBar";

describe("<Menu>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Menu />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Menu />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const wrapper = mount(<Menu />);

      expect(wrapper.find("Menu").props().closeOnSelect).toEqual(true);
      expect(wrapper.find("Menu").props().isOpen).toEqual(true);
      expect(wrapper.find("Menu").props().keepOnScreen).toEqual(false);
      expect(wrapper.find("Menu").props().popoverStyles).toEqual({});
    });
  });

  describe("anchorEl prop", () => {
    it("passes anchorEl prop to correct child", () => {
      const anchorEl = ReactTestUtils.renderIntoDocument(<AppBar />);

      const wrapper = mount(<Menu anchorEl={anchorEl} />);

      expect(wrapper.find("Popover").props().anchorEl).toEqual(anchorEl);
    });
  });

  describe("anchorPosition prop", () => {
    it("passes anchorPosition prop to correct child", () => {
      const anchorPosition = { horizontal: "left", vertical: "bottom" };

      const wrapper = mount(<Menu anchorPosition={anchorPosition} />);

      expect(wrapper.find("Popover").props().anchorPosition).toEqual(
        anchorPosition
      );
    });
  });

  describe("children prop", () => {
    it("renders given children", () => {
      const child = <div>child</div>;
      const anchorEl = ReactTestUtils.renderIntoDocument(
        <div>
          <AppBar />
        </div>
      );

      const wrapper = mount(<Menu anchorEl={anchorEl}>{child}</Menu>);

      expect(wrapper.containsMatchingElement(child)).toBeTruthy();
    });
  });

  describe("classes prop", () => {
    it("sets custom class", () => {
      const classes = ["newClass"];
      const anchorEl = ReactTestUtils.renderIntoDocument(
        <div>
          <AppBar />
        </div>
      );

      const wrapper = mount(
        <Menu anchorEl={anchorEl} classes={classes}>
          <MenuItem />
        </Menu>
      );

      expect(
        wrapper
          .find("MenuList")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClasses decorator", () => {
      const classes = ["riipen", "riipen-menu"];

      const wrapper = mount(<Menu />);

      expect(
        wrapper
          .find("Menu")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });
  });

  describe("closeOnSelect prop", () => {
    it("does not invoke onClose when closeOnSelect is false", () => {
      const closeOnSelect = false;
      const onClose = jest.fn();
      const anchorEl = ReactTestUtils.renderIntoDocument(
        <div>
          <AppBar />
        </div>
      );

      const wrapper = mount(
        <Menu
          anchorEl={anchorEl}
          closeOnSelect={closeOnSelect}
          onClose={onClose}
        >
          <MenuItem />
        </Menu>
      );

      wrapper
        .find("MenuItem")
        .childAt(0)
        .simulate("click");

      expect(onClose).toHaveBeenCalledTimes(0);
    });
  });

  describe("color prop", () => {
    it("passes color prop to correct child", () => {
      const color = "secondary";
      const anchorEl = ReactTestUtils.renderIntoDocument(
        <div>
          <AppBar />
        </div>
      );

      const wrapper = mount(
        <Menu anchorEl={anchorEl} color={color}>
          <MenuItem />
        </Menu>
      );

      expect(wrapper.find("MenuList").props().color).toEqual(color);
    });

    it("gives an error when given an invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Menu color="blue" />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("contentPosition prop", () => {
    it("passes contentPosition prop to correct child", () => {
      const contentPosition = { horizontal: "left", vertical: "bottom" };

      const wrapper = mount(<Menu contentPosition={contentPosition} />);

      expect(wrapper.find("Popover").props().contentPosition).toEqual(
        contentPosition
      );
    });
  });

  describe("fullWidth prop", () => {
    it("passes fullWidth prop to correct child", () => {
      const fullWidth = true;

      const wrapper = mount(<Menu fullWidth={fullWidth} />);

      expect(wrapper.find("Popover").props().fullWidth).toEqual(fullWidth);
    });
  });

  describe("isOpen prop", () => {
    it("passes isOpen prop to correct child", () => {
      const isOpen = true;

      const wrapper = mount(<Menu isOpen={isOpen} />);

      expect(wrapper.find("Popover").props().isOpen).toEqual(isOpen);
    });
  });

  describe("keepOnScreen prop", () => {
    it("passes keepOnScreen prop to correct child", () => {
      const keepOnScreen = true;

      const wrapper = mount(<Menu keepOnScreen={keepOnScreen} />);

      expect(wrapper.find("Popover").props().keepOnScreen).toEqual(
        keepOnScreen
      );
    });
  });

  describe("marginThreshold prop", () => {
    it("passes marginThreshold prop to correct child", () => {
      const marginThreshold = 2;

      const wrapper = mount(<Menu marginThreshold={marginThreshold} />);

      expect(wrapper.find("Popover").props().marginThreshold).toEqual(
        marginThreshold
      );
    });
  });

  describe("onClose prop", () => {
    it("invokes onClose when child is clicked", () => {
      const onClose = jest.fn();
      const anchorEl = ReactTestUtils.renderIntoDocument(
        <div>
          <AppBar />
        </div>
      );

      const wrapper = mount(
        <Menu anchorEl={anchorEl} onClose={onClose}>
          <MenuItem />
        </Menu>
      );

      wrapper
        .find("MenuItem")
        .childAt(0)
        .simulate("click");

      expect(onClose).toHaveBeenCalledTimes(1);
    });

    it("does not invoke onClose when event type is not a click event", () => {
      const onClose = jest.fn();
      const anchorEl = ReactTestUtils.renderIntoDocument(
        <div>
          <AppBar />
        </div>
      );

      const wrapper = mount(
        <Menu anchorEl={anchorEl} onClose={onClose}>
          <MenuItem />
        </Menu>
      );

      wrapper
        .find("MenuItem")
        .childAt(0)
        .simulate("submit");

      expect(onClose).toHaveBeenCalledTimes(0);
    });
  });

  describe("onSelect prop", () => {
    it("invokes onSelect when child is clicked", () => {
      const onSelect = jest.fn();
      const anchorEl = ReactTestUtils.renderIntoDocument(
        <div>
          <AppBar />
        </div>
      );

      const wrapper = mount(
        <Menu anchorEl={anchorEl} onSelect={onSelect}>
          <MenuItem />
        </Menu>
      );

      wrapper
        .find("MenuItem")
        .childAt(0)
        .simulate("click");

      expect(onSelect).toHaveBeenCalledWith(0, expect.any(Object));
    });
  });

  describe("popoverStyles prop", () => {
    it("passes popoverStyles prop to correct child", () => {
      const styles = ReactTestUtils.renderIntoDocument(<style />);

      const wrapper = mount(<Menu popoverStyles={styles} />);

      expect(wrapper.find("Popover").props().styles).toEqual(styles);
    });
  });

  describe("selectedIndex prop", () => {
    it("passes selectedIndex prop to correct child", () => {
      const selectedIndex = 0;
      const anchorEl = ReactTestUtils.renderIntoDocument(
        <div>
          <AppBar />
        </div>
      );

      const wrapper = mount(
        <Menu anchorEl={anchorEl} selectedIndex={selectedIndex}>
          <MenuItem />
        </Menu>
      );

      expect(wrapper.find("MenuList").props().selectedIndex).toEqual(
        selectedIndex
      );
    });
  });
});
