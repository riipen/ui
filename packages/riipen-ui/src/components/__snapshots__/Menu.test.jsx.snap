// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu> renders correct snapshot 1`] = `
<withClasses(Menu)>
  <Menu
    classes={
      Array [
        "riipen",
        "riipen-menu",
      ]
    }
    closeOnClick={true}
    color="primary"
    isOpen={true}
    keepOnScreen={false}
    popoverStyles={Object {}}
    variant="menu"
  >
    <withClasses(Popover)
      isOpen={true}
      keepOnScreen={false}
      lockScroll={false}
      styles={Object {}}
    >
      <Popover
        classes={
          Array [
            "riipen",
            "riipen-popover",
          ]
        }
        closeOnScrolled={true}
        component="span"
        fullWidth={false}
        isOpen={true}
        keepOnScreen={false}
        lockScroll={false}
        marginThreshold={16}
        styles={Object {}}
      >
        <JSXStyle
          dynamic={
            Array [
              "#fff",
              "4px",
              "0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12)",
              "calc(100% - 32px)",
              "",
              3,
            ]
          }
          id="631859544"
        />
      </Popover>
    </withClasses(Popover)>
  </Menu>
</withClasses(Menu)>
`;
